sdk/clientcore/annotation-processor/src/main/java/io/clientcore/annotation/processor/utils/ResponseHandler.java

```diff
@@ -40,50 +43,98 @@ public final class ResponseHandler {
+
+        body.tryAddImportToParentCompilationUnit(HttpResponseException.class);
+        BlockStmt ifBlock = new BlockStmt();
+        ifBlock.addStatement(
+            StaticJavaParser.parseStatement("String errorMessage = networkResponse.getValue().toString();"));
+        if (!usingTryWithResources) {
+            closeResponse(ifBlock);
+        }
+        ifBlock.addStatement(
+            StaticJavaParser.parseStatement("throw new HttpResponseException(errorMessage, networkResponse, null);"));
+        IfStmt ifStmt = new IfStmt()
+            .setCondition(new UnaryExpr(new NameExpr("expectedResponse"), UnaryExpr.Operator.LOGICAL_COMPLEMENT))
+            .setThenStmt(ifBlock);
+        body.addStatement(ifStmt);
+    }
+
+    private static void handleRequestReturn(BlockStmt body, TypeMirror returnType, java.lang.reflect.Type entityType,
+        HttpRequestContext method, boolean serializationFormatSet) {
         boolean returnIsResponse = TypeConverter.isResponseType(returnType);
 
         // TODO (alzimmer): Base64Uri needs to be handled. Determine how this will show up in code generation and then
         //  add support for it.
-        if (entityType == Void.TYPE || entityType == Void.class) {
-            // Return is void, Void, or Response<Void>, close the network response and return null as the value.
+        if (returnType.getKind() == TypeKind.VOID) {
             closeResponse(body);
+        } else if (entityType == Void.TYPE || entityType == Void.class) {
             addReturnStatement(body, returnIsResponse, "null");
         } else if (method.getHttpMethod() == HttpMethod.HEAD && isBooleanType(entityType)) {
             // HTTP method was either HEAD or the return is a boolean. Use the status code to determine response value.
-            // Always close the network response in this situation as we'll be throwing it away.
-            closeResponse(body);
             addReturnStatement(body, returnIsResponse, "expectedResponse");
         } else if (TypeUtil.isTypeOrSubTypeOf(entityType, byte[].class)) {
             // Return is a byte[]. Convert the network response body into a byte[].
             body.addStatement(StaticJavaParser.parseStatement("BinaryData responseBody = networkResponse.getValue();"));
-            body.addStatement(StaticJavaParser
-                .parseStatement("byte[] responseBodyBytes = responseBody != null ? responseBody.toBytes() : null;"));
-
-            // Always close the network response in this situation as converting to a byte[] will read the entire
-            // response. This should complete the connection, but let's be safe. This must be done after getting the
-            // byte[] though.
-            closeResponse(body, "Close the network response as the body should be consumed.");
 
             // Return responseBodyBytes as-is which will have the behavior of null -> null, empty -> empty, and
```

Comment:
nit; `responseBodyBytes` is not used as a variable any longer.

Reply:
Updated comment